scripts/install/RegisterCounters.php

```
@@ -0,0 +1,68 @@
+use oat\oatbox\service\ServiceNotFoundException;
+use oat\tao\model\counter\CounterService;
+use oat\tao\model\counter\CounterServiceException;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use oat\taoDelivery\model\execution\StateService;
+use oat\taoDelivery\model\execution\StateServiceInterface;
+use oat\taoDelivery\models\classes\execution\event\DeliveryExecutionCreated;
+use oat\taoDelivery\models\classes\execution\event\DeliveryExecutionState;
+
+/**
+ * Script registers listeners for delivery execution event which counts related actions.
+ *
+ * Script usage: sudo -u www-data php index.php 'oat\taoDelivery\scripts\install\RegisterCounters'
+ */
+class RegisterCounters extends InstallAction
+{
+    public const COUNTER_SHORT_NAME_DELIVERY_EXECUTION_CREATED = 'taoDelivery:created';
+    public const COUNTER_SHORT_NAME_DELIVERY_EXECUTION_STATE = 'taoDelivery:state';
+
+    /**
+     * @param array $params
+     * @throws common_Exception
+     * @throws ServiceNotFoundException
+     * @throws InvalidServiceManagerException
+     * @throws CounterServiceException
+     */
+    public function __invoke($params = [])
+    {
+        /** @var CounterService $counterService */
+        $counterService = $this->getServiceManager()->get(CounterService::SERVICE_ID);
+
+        // Attach to Delivery Execution Creation Event.
+        $counterService->attach(
+            DeliveryExecutionCreated::class,
+            self::COUNTER_SHORT_NAME_DELIVERY_EXECUTION_CREATED
+        );
+
+        // Attach to Delivery Execution State Event.
+        // The DeliveryExecutionState::getState() method is used to count depending on the
+        // returned value.
+        $counterService->attach(
+            DeliveryExecutionState::class,
+            self::COUNTER_SHORT_NAME_DELIVERY_EXECUTION_STATE,
+            'getHumanReadableState'
+        );
+
+        // Resetting counters and setting starting values for SQL based KV storages
```

Same here. Why this comment is necessary since you call `$counterService->reset`? Another thing is that, if you change your storage from KV to something else, this comment is already outdated.

---

The issue here is that we already have several implementations of KV, based on different services. In case of Redis no need to reset values, because incrementing of an empty value sets it to 0 first, and after just adds 1 to the 0. So everything works fine. But in the case of KV over SQL, increment does not work, because the initial value is null (or even does not exists as an entry), and simple incrementing tries to add 1 to null, and postgress fails such request. Maybe I have to rephrase the comment, but I tried to explain why it was needed here.

---

So you can still remove this comment and throw an exception on `CounterService` in case an invalid persistence is provided. Wdyt? Cause not sure a comment will solve the issue :)

---

But it is not about invalid persistence, it is about valid persistence, which has some limitations (not true KV storage).
For invalid storage already exists proper exceptions and it works. Comment here with the purpose to explain unobvious things. I don't think that we have to avoid use comments at all in the code. I will leave a comment but update it with a proper explanation.

---

I am assuming then all places that we call `$counterService->reset` we have to write a comment?

---

Nope, it is required only once when a new counter listener is attached and configured. Another solution to check if the key already exists in the `increment` method, every method call. I think you will agree that it is worse.

---

I mean, this is a public method. We cannot predict all the places and how will it be used. Why not make it private then an call it inside `CounterService::attach`? Since it seems mandatory

---

It should be public because it also can be used for resetting the values to 0, I mean just cleat the counters. One of the AC requires possibility to do that once a day, at night. For it, I am going to create a special script, which will utilize this method. The logic is clear, you are setting the counter, then resetting it, and starting count. Latter if needed you can reset the counter by the public method based on conditions.

---

As a solution we can do both. Leave it publish to allow manual reset and call reset method inside attach to make sure we set initial value to 0 in SQL KV persistence implementations, it will be ensure that no one will register counter and forget to call reset. WDYT?

---

I also wanted to do that, but trick there, that service can call the callback, which returns the value used for a key. Like with `getState`, when it can return `Finished`, or `Active`. In case, when I have to reset it inside `attach`, I need to know what can be returned, but it is not the responsibility of the counter service, it is something outside and it comes from the configuration.